WordPress/gutenberg
# Try: extract strings for translation directly from the source JSON file

This is an experiment to showcase an alternative approach to handle the translation of strings stored in JSON files.

Related:

- `block.json` https://github.com/wp-cli/i18n-command/pull/210
- `theme.json` https://github.com/wp-cli/i18n-command/issues/224

https://github.com/WordPress/gutenberg/pull/27380 introduced a way for extracting strings from theme.json for translation, based on the idea that having a shared [intermediate file](https://github.com/wp-cli/i18n-command/issues/224#issuecomment-731122437) for core, gutenberg, and wp-cli would allow updating the theme.json structure without requiring changes to the code that process those strings.

Initially, the aim was to translate only presets. However, `theme.json` aims now to absorb more theme metadata that may need to be translated https://github.com/WordPress/gutenberg/pull/28110 and information for translator context was added recently as well https://github.com/WordPress/gutenberg/pull/28246 While working on adding support for translating the `customTemplates` titles at https://github.com/WordPress/gutenberg/pull/28783 I found the current direction too complex to implement with the current theme.json shape.

This PR proposes an alternative: what if we marked the strings to translate directly in the source JSON file (block.json, theme.json) instead of having intermediate structures (code, i18.json file)? For familiarity, I used `_x( string-to-translate, context-for-translators )` but any mark could be used.

For example, this would be a theme.json in which the color preset names and other fields are marked for translation:

```json
{
  "settings": {
    "defaults": {
      "color": {
        "palette": [
          {
            "slug": "black",
            "color": "#000000",
            "name": "_x( Black, Color name )"
          }
        ]
      }
    }
  },
  "customTemplates": {
    "page-home": {
      "title": "_x( Page title, Context for this string )"
    }
  }
}
```

This PR provides a prototype that implements this direction:

- In terms of code maintainability, it's simpler: [code added](https://github.com/WordPress/gutenberg/commit/5b1e57aec35147a00fc4b642020e3d091c04e4ee), [code removed](https://github.com/WordPress/gutenberg/commit/0ba281bcc65f33f9dbe9a4b9cbd440cdb642660a).
- It doesn't require any code change as those JSON structures evolve, hence it scales well to any JSON structure: can be used for any file (theme.json, block.json, etc) and field (no matter the shape of the file).

cc @swissspidy @schlessera @gziolo for thoughts.

## How to test

- Look at the `test_get_template_parts` test.
- Alternatively, using this PR from TT1-blocks, load any editor and verify that the UI color controls still show the proper names for "Black" and "Dark Gray". https://github.com/WordPress/theme-experiments/pull/206
